### PR TITLE
Provide a way to disable `DnsQueryLifecycleObserverFactory`

### LINKED ISSUE #3368: Provide a way to disable the default DNS metrics
Some users might not want to monitor DNS metrics because the metrics could contain various domains as metric keys if a client sends requests to unspecified domains such as webhook domains or redirect locations.

As a workaround, the metics could be filtered by [meter filters](https://armeria.dev/docs/advanced-metrics/#excluding-certain-meters-created-by-armeria).
However, it would be better to disable the metrics if they don't need them.
For example:
```java
DnsResolverGroupBuilder.disableDnsQueryLifecycleObserverFactory()
```

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Some users might not want to monitor DNS metrics
if a client sends requests to unspecified domains
such as webhook domains or redirect locations.
In that case, the cardinality of the metrics could be huge
because of various domain names.

Modifications:

- Add `disableDnsQueryLifecycleObserverFactory()` option to `DnsResolverGroupBuilder`

Result:

- You can now disable DNS metrics using `DnsResolverGroupBuilder.disableDnsQueryLifecycleObserverFactory()`.
  ```java
  ClientFactory.builder()
               .domainNameResolverCustomizer(builder -> {
                   builder.disableDnsQueryLifecycleObserverFactory();
               })
  ```
- Fixes #3368

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java [hunk 1]
```diff
@@ -361,13 +376,17 @@ RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
                     new DefaultDnsQueryLifecycleObserverFactory(
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:376]: `observerFactory` is not used when `disableDnsQueryLifecycleObserverFactory` is false so we should remove it. 😄

File: core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java [hunk 2]
```diff
@@ -361,13 +376,17 @@ RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
...
                             meterRegistry,
                             new MeterIdPrefix("armeria.client.dns.queries"));
-            if (dnsQueryLifecycleObserverFactory == null) {
-                builder.dnsQueryLifecycleObserverFactory(observerFactory);
-            } else {
-                builder.dnsQueryLifecycleObserverFactory(
-                        new BiDnsQueryLifecycleObserverFactory(observerFactory,
-                                                               dnsQueryLifecycleObserverFactory));
+
+            if (!disableDnsQueryLifecycleObserverFactory) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:380]: I guess there might be a chance that a user wants to set a customized `DnsQueryLifecycleObserver` while he/she doesn't want to use the `DefaultDnsQueryLifecycleObserverFactory`.
So how about changing the parameter to `disableDnsQueryMetric` and allow users to use the customized factory?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:380]: +1 Nice idea.

File: core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java [hunk 3]
```diff
@@ -324,6 +324,53 @@ void cname() {
...
+                assertThat(MoreMeters.measureAll(meterRegistry))
+                        .doesNotContainKeys(writeMeterId, successMeterId);
+
+                client.get("http://foo.com:1/").aggregate();
+
+                // Give enough time for metrics to be collected by the MeterRegistry
+                Thread.sleep(1000);
+                assertThat(MoreMeters.measureAll(meterRegistry).entrySet().stream()
+                                     .anyMatch(entry -> entry.getKey()
+                                                             .contains("armeria.client.dns.queries")))
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java:368]:
```suggestion
                                                             .startsWith("armeria.client.dns.")))
```

File: core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java [hunk 4]
```diff
@@ -324,6 +324,53 @@ void cname() {
...
+                final String writeMeterId =
+                        "armeria.client.dns.queries.written#count{name=foo.com.,server=" +
+                        getHostAddress(server) + '}';
+                final String successMeterId =
+                        "armeria.client.dns.queries#count{cause=none,name=foo.com.,result=success}";
+                final String otherExceptionId =
+                        "armeria.client.dns.queries#count{" +
+                        "cause=others,name=bar.com.,result=failure}";
+                assertThat(MoreMeters.measureAll(meterRegistry))
+                        .doesNotContainKeys(writeMeterId, successMeterId);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java:360]: How about making sure there are no meters that start with `armeria.client.dns.`?

File: core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java [hunk 5]
```diff
@@ -90,6 +90,7 @@
     private DnsServerAddressStreamProvider dnsServerAddressStreamProvider;
     @Nullable
     private DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory;
+    private boolean disableDnsQueryMetric;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:93]: `dnsQueryMetricsDisabled`?

File: core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java [hunk 6]
```diff
@@ -261,6 +262,15 @@ public DnsResolverGroupBuilder dnsQueryLifecycleObserverFactory(
         return this;
     }
 
+    /**
+     * Disables the default {@link DnsQueryLifecycleObserverFactory} that collects DNS query metrics through
+     * {@link MeterRegistry}.
+     */
+    public DnsResolverGroupBuilder disableDnsQueryMetric() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:269]: `disableDnsQueryMetrics` (s at the end)

**Review by minwoox — APPROVED** [review]
LGTM once @trustin's comments are addressed. 😄

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon !

**Pull request merged**